staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/validation_test.go

```diff
@@ -23,6 +23,80 @@ import (
 	fuzz "github.com/google/gofuzz"
 )
 
+func TestValidateStructuralInvariantsMetadata(t *testing.T) {
+	fuzzer := fuzz.New()
+	fuzzer.Funcs(
+		func(s *JSON, c fuzz.Continue) {
+			if c.RandBool() {
+				s.Object = float64(42.0)
+			}
+		},
+		func(s **StructuralOrBool, c fuzz.Continue) {
+			if c.RandBool() {
+				*s = &StructuralOrBool{}
+			}
+		},
+		func(s **Structural, c fuzz.Continue) {
+			if c.RandBool() {
+				*s = &Structural{}
+			}
+		},
+		func(s *Structural, c fuzz.Continue) {
+			if c.RandBool() {
+				*s = Structural{}
+			}
+		},
+		func(vv **ValueValidation, c fuzz.Continue) {
+			if c.RandBool() {
+				*vv = &ValueValidation{}
+			}
+		},
+	)
+	fuzzer.NilChance(0).NumElements(1,1)
+
+	// check that anything other than name and generateName in metadata properties is forbidden
+	properties := []string {"name", "generateName", "testProperty"}
+	for _, property := range properties {
+		s := &Structural{
+			Properties: map[string]Structural{
+				"metadata": {
+					Properties: map[string]Structural{
+						property: {},
+					},
+				},
+			},
+		}
+		errs := validateStructuralInvariants(s, rootLevel, nil)
+		if len(errs) != 0 {
+			t.Logf("errors returned: %v", errs)
+		}
+		if len(errs) == 0 && property != "name" && property != "generateName" {
+			t.Errorf("expected forbidden field validation errors for: %#v", s)
+		}
```

Comment:
you don't test that name and generateName are allowed.

Reply:
@sttts Allowed cases will bring many additional mocks for `Structural` because we have to ensure the object will pass other validations, so can we split the validation for metadata to a new function like `validateStructuralMetadata`? It can reduce lots of redundant works here.

Reply:
Sure, if it helps move the tests to `validateStructuralMetadataInvariants`.

My point here is that I want to see that there are no other errors in `errs` that make `s` forbidden. Hence, showing that `name` and `generateName` here lead to `len(errs)==0` would show that.